### ISSUE TITLE
allow positional parameters in template strings

### ISSUE THREAD

**Issue by aoisensi** [odota/web#627]
https://github.com/odota/ui/commit/c0acd3d002a43e1e716a8c923541876fdf3cd540

Thank you for try to support formatting text.  
But, `utils.format` is not support choice order.  
(like `printf("%2$d", 1, 2) // =>2`)  
Need to use some format library.

**Comment by howardchung** [comment]
Is it possible to phrase the translation in such a way that it does not require changing parameter order?

**Comment by aoisensi** [comment]
I can. But so much unnative Japanese.

**Comment by howardchung** [comment]
https://www.npmjs.com/package/string-template probably works for this.  I'd prefer not to add additional dependencies if possible

**Comment by aoisensi** [comment]
hmm..
However, almost translator want localize package.

**Comment by howardchung** [comment]
update:

We can probably reuse the function for the story that takes named parameters in translation strings (and accepts a object with key-value pairs for the parameters)

**Comment by gu3st** [comment]
I'll look to work on this issue this coming weekend.  From our Discord conversations, I might look to extend the string interpolation strategy that exists in `/src/utility/index.jsx#207` to introduce indexed arguments (possibly the :formatStr arg that C# implements as well). 

I'd change everything to use formatTemplate, but converting everywhere that we use util.format to utilize an object, and then update the corresponding 30+ translation files is a fairly large task to undertake.  I'll have to see how ambitious I am.  

**TL;DR**:

1. Update formatTemplate to accept {<index>:<format>} as a template string identifier

2. Update formatTemplate to accept variable amounts of arguments.  If 2 arguments and 2nd arg is Object, assume it's the same template/dict structure.  if >=2 args and 2nd arg is string/number, assume it's similar to util.format.

3. Update translation files to utilize {<num>} instead of %s format strings.  This should be able to be done via a migration script.

Does this all sound sane?  Let me know if I should tackle it differently or if there's any feedback.

**Comment by howardchung** [comment]
can you give an example of 1)?

**Comment by howardchung** [comment]
I see a couple examples of numeric parameters already, perhaps we support this in some way already?

https://github.com/odota/web/blob/master/src/lang/en-US.json#L640

@mdiller might know the specifics

**Comment by howardchung** [comment]
also while we're on this topic:

When @albertcui implemented the APi documentation he picked a different syntax for injecting parameters, prefixing the term with `$`.  We might want to pick one convention and stick with it:

https://github.com/odota/web/blob/master/src/lang/en-US.json#L50

**Comment by gu3st** [comment]
https://github.com/odota/web/blob/master/src/components/Match/MatchStory.jsx#L169

It looks like an object is being passed in with numerical keys.

As for 1), Formatted strings usually allow you to pass some type with them to get converted to.  C# accepts similar (and more).  I was just thinking to allow you to pass at least some of the basic types to allow you to parseInt/parseFloat which is expected behaviour in most string formatters.

**Comment by gu3st** [comment]
So this is what I have so far: 

https://github.com/odota/web/compare/arbitrary-position-template-strings

I updated the formatTemplate function so it checks if the dict argument is an object. If so it retains the old behaviour.  If not, it automatically converts the argument list to a numbered object, as to work with `{1}` style template strings.   I also added a 2nd function that calls .join('') on the result of formatTemplate. I'd have made the string conversion automatic on formatTemplate but Stories relies on the ability to template components into a token.

I also updated all of the translations via a script.  It's a fairly simple script that reads for any %[char] and converts it to {idx} in all the translations. It takes a backup as well but I did not commit these to the repository (git history works for that).

The only unknown right now (to me) is updating the translations with the migrated translations on https://crowdin.com/translate/yasp/ which I think would be the last step.  I QAed each place where I changed `util.format` to `formatTemplate` with English locale at least.  I didn't audit all the languages to see if there were any errors/mistakes.

I'll create a pull request if all of this seems sane, or if there's more things for me to change/work on.
